src/app/templates/SwapForm.tsx

```diff
@@ -279,39 +267,80 @@ const SwapForm: React.FC<SwapFormProps> = ({ defaultAsset }) => {
     }
     const rawExchangeRate = inputAssetAmount.div(outputAssetAmount);
     if (rawExchangeRate.eq(0)) {
-      return { base: new BigNumber(1), value: new BigNumber(0) };
+      return {base: new BigNumber(1), value: new BigNumber(0)};
     }
     const base = new BigNumber(10).pow(
-      BigNumber.max(
-        0,
-        -Math.floor(Math.log10(rawExchangeRate.toNumber())) -
-          inputAsset.decimals
-      )
+        BigNumber.max(
+            0,
+            -Math.floor(Math.log10(rawExchangeRate.toNumber())) -
+            inputAsset.decimals
+        )
     );
     const prettifiedExchangeRate = rawExchangeRate
-      .multipliedBy(base)
-      .decimalPlaces(inputAsset.decimals);
-    return { base, value: prettifiedExchangeRate };
+        .multipliedBy(base)
+        .decimalPlaces(inputAsset.decimals);
+    return {base, value: prettifiedExchangeRate};
   }, [inputAssetAmount, outputAssetAmount, inputAsset]);
 
+  const inputContractAddress = useMemo(() => {
+    if (!inputAsset) {
+      return undefined;
+    }
+    return getAssetExchangeData(
+        tokensExchangeData,
+        tezUsdPrice,
+        inputAsset!,
+        selectedExchanger
+    )?.contract;
+  }, [tokensExchangeData, tezUsdPrice, inputAsset, selectedExchanger])
+
+  const outputContractAddress = useMemo(() => {
+    if (!outputAsset) {
+      return undefined;
+    }
+    return getAssetExchangeData(
+        tokensExchangeData,
+        tezUsdPrice,
+        outputAsset!,
+        selectedExchanger
+    )?.contract;
+  }, [tokensExchangeData, tezUsdPrice, outputAsset, selectedExchanger]);
+
+  const feePercentage = useMemo(() => getFeePercentage(selectedExchanger), [selectedExchanger]);
+
+  const priceImpact = usePriceImpact(tezos, selectedExchanger, inputContractAddress, outputContractAddress, inputAssetAmount, outputAssetAmount, feePercentage, inputAsset, outputAsset);
+
+  const getFormattedPriceImpact = (priceImpact: BigNumber) => {
+    if (priceImpact.toString() !== "0"
+        && priceImpact.toString() !== "NaN"
+        && priceImpact.toString() !== "Infinity"
+        && priceImpact.toString() !== "-Infinity") {
+
+      return priceImpact.toFixed(2) + "%";
+    } else {
+      return "-";
+    }
+  }
```

Comment:
this fn should not be re-created on each rerender.
move it put of the compknent into seperate file